### PR TITLE
Refactor chains/nsmgrproxy to use Option pattern

### LINKED ISSUE #736: Refactor chains/nsmgrproxy to use Option pattern
## Motivation

We can simplify using of our chains with
[Option pattern](https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md)

See at an example: https://github.com/networkservicemesh/sdk/commit/9007c4530373d301993ecd8985a1c9abebe359b2

Moreover, this can super help us with minimization of manual changes of upstream repositories.

Entry point: https://github.com/networkservicemesh/sdk/blob/master/pkg/networkservice/chains/nsmgrproxy/server.go

### PR DESCRIPTION
**Pull request by Tiberivs**
Refactor `chains/nsmgrproxy` to use Option pattern
Closes #736.

Signed-off-by: Anatoly Loskutnikov <anatoly.loskutnikov@gmail.com>

### CODE REVIEW

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 1]
```diff
@@ -22,30 +22,78 @@ import (
...
+	return func(o *serverOptions) {
+		o.grpcOptions = options
+	}
+}
+
 // NewServer creates new proxy NSMgr
-func NewServer(ctx context.Context, name string, authorizeServer networkservice.NetworkServiceServer, tokenGenerator token.GeneratorFunc, options ...grpc.DialOption) endpoint.Endpoint {
+func NewServer(ctx context.Context, options ...Option) endpoint.Endpoint {
+	opts := &serverOptions{
+		name:            "server-" + uuid.New().String(),
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:79]:
```suggestion
		name:            "nsmgr-proxy-" + uuid.New().String(),
```

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 2]
```diff
@@ -22,30 +22,78 @@ import (
...
+
+// WithTokenGenerator sets token generator function for the server
+func WithTokenGenerator(function token.GeneratorFunc) Option {
+	return func(o *serverOptions) {
+		o.tokenGenerator = function
+	}
+}
+
+// WithGRPCOptions sets gRPC Dial Options for the server
+func WithGRPCOptions(options ...grpc.DialOption) Option {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:70]: Please follow the steps from the issue description.

```suggestion
func WithDialOption(options ...grpc.DialOption) Option {
```

File: pkg/tools/sandbox/builder.go [hunk 3]
```diff
@@ -188,7 +188,11 @@ func (b *Builder) newNSMgrProxy(ctx context.Context) *EndpointEntry {
 		return nil
 	}
 	name := "nsmgr-proxy-" + uuid.New().String()
-	mgr := b.supplyNSMgrProxy(ctx, name, authorize.NewServer(authorize.Any()), b.generateTokenFunc, DefaultDialOptions(b.generateTokenFunc)...)
+	mgr := b.supplyNSMgrProxy(ctx,
+		nsmgrproxy.WithName(name),
+		nsmgrproxy.WithAuthorizeServer(authorize.NewServer(authorize.Any())),
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/builder.go:193]:
```suggestion
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 4]
```diff
@@ -22,30 +22,78 @@ import (
...
+		o.authorizeServer = s
+	}
+}
+
+// WithTokenGenerator sets token generator function for the server
+func WithTokenGenerator(function token.GeneratorFunc) Option {
+	return func(o *serverOptions) {
+		o.tokenGenerator = function
+	}
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:67]: This is required parameter... Please see at Option pattern description in the Issue.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/nsmgrproxy/server.go [hunk 5]
```diff
@@ -22,30 +22,74 @@ import (
...
+	opts := &serverOptions{
+		name:            "nsmgr-proxy-" + uuid.New().String(),
+		authorizeServer: authorize.NewServer(authorize.Any()),
+	}
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	if opts.authorizeServer == nil {
+		panic("Authorize server is nil")
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server.go:79]: Why do not panic at line 57?

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 1 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: Renamed

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 2 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: Done

File: pkg/tools/sandbox/builder.go (hunk 3 above)
Comment (Tiberivs) [reply pkg/tools/sandbox/builder.go]: Removed

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 4 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: Removed from options

File: pkg/networkservice/chains/nsmgrproxy/server.go (hunk 5 above)
Comment (Tiberivs) [reply pkg/networkservice/chains/nsmgrproxy/server.go]: Done

**Pull request merged**